### PR TITLE
Make REST Bundle optional

### PR DESCRIPTION
**Pull request by core23**
As the REST bundle is not required for the this bundle, the api form and controllers definitions should only be loaded, when the bundles are used.

### CODE REVIEW

**Comment by core23** [comment]
As the REST bundle is not required for the this bundle, the api form and controllers definitions should only be loaded, when the bundles are used.

File: DependencyInjection/SonataMediaExtension.php [hunk 1]
```diff
@@ -53,16 +53,20 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('validators.xml');
         $loader->load('serializer.xml');
 
-        if (!in_array(strtolower($config['db_driver']), array('doctrine_orm', 'doctrine_mongodb', 'doctrine_phpcr'))) {
-            throw new \InvalidArgumentException(sprintf('SonataMediaBundle - Invalid db driver "%s".', $config['db_driver']));
-        }
+        $bundles = $container->getParameter('kernel.bundles');
+
+        if (isset($bundles['FOSRestBundle']) && isset($bundles['NelmioApiDocBundle'])) {
```
Comment (Soullivaneuh) [inline DependencyInjection/SonataMediaExtension.php]: Not sure that `NelmioApiDocBundle` is required to get REST working.

**Comment by core23** [comment]
I copied the detection from the NewsBundle. Thought this would be the same logic

**Comment by core23** [comment]
@rande same problem like https://github.com/sonata-project/SonataClassificationBundle/pull/42

**Comment by core23** [comment]
ping @rande

**Comment by core23** [comment]
Sorry for bothering you, but can you have a quick look? @rande

**Comment by core23** [comment]
This bug is really annoying... Can you have a look @Soullivaneuh as @rande seems a little busy.

File: DependencyInjection/SonataMediaExtension.php [hunk 2]
```diff
@@ -51,16 +51,18 @@ public function load(array $configs, ContainerBuilder $container)
...
         $loader->load('serializer.xml');
 
-        if (!in_array(strtolower($config['db_driver']), array('doctrine_orm', 'doctrine_mongodb', 'doctrine_phpcr'))) {
-            throw new \InvalidArgumentException(sprintf('SonataMediaBundle - Invalid db driver "%s".', $config['db_driver']));
-        }
+        $bundles = $container->getParameter('kernel.bundles');
 
-        $loader->load(sprintf('api_form_%s.xml', $config['db_driver']));
+        if (isset($bundles['FOSRestBundle']) && isset($bundles['NelmioApiDocBundle'])) {
+            if (!in_array(strtolower($config['db_driver']), array('doctrine_orm', 'doctrine_mongodb', 'doctrine_phpcr'))) {
```
Comment (rande) [inline DependencyInjection/SonataMediaExtension.php]: Why did you move this code ? seems to be unrelated to bundles detections.

Comment (core23) [inline DependencyInjection/SonataMediaExtension.php]: Thanks for your advice. I moved it to the origin @rande

**Comment by OskarStark** [comment]
@core23 it is not optional IMO, see: #877 

we need this use statements and have the API Doc there, so i'm :+1: for closing this PR

**Comment by core23** [comment]
It IS optional, as I added a check to load the bundles only when the API bundles are available.

**Comment by OskarStark** [comment]
ah ok, sorry you are right, i was confused :)

File: composer.json [hunk 3]
```diff
@@ -40,7 +38,9 @@
         "aws/aws-sdk-php": "~2.7",
         "doctrine/mongodb-odm": "~1.0",
         "jackalope/jackalope-doctrine-dbal": "~1.1",
-        "symfony/phpunit-bridge": "~2.7|~3.0"
+        "symfony/phpunit-bridge": "~2.7|~3.0",
+        "friendsofsymfony/rest-bundle": "~1.1",
```
Comment (OskarStark) [inline composer.json]: are you sure, that ``require-dev`` ist correct? not ``suggest``?

Comment (core23) [inline composer.json]: They are required(-dev) for the travis build.

Comment (OskarStark) [inline composer.json]: hmm, but for the development it looks weird, so on my local machine a would have an api with docs and on prod not... :-1:

Comment (core23) [inline composer.json]: Have a look at the composer docs (https://getcomposer.org/doc/04-schema.md#require-dev)

The dev packages are only loaded if you have them in your root ``composer.json``. No package of a child ``composer.json`` is loaded.

**Comment by core23** [comment]
Any news for this bug? @rande @OskarStark

**Comment by OskarStark** [comment]
I'll let @rande decide

**Comment by core23** [comment]
Sorry for pinging you that much, but that needless dependency is really annoying... @rande

**Comment by core23** [comment]
Thank you :heart: @rande
